src/abstract-interpretation/absint-visitor.ts

```diff
@@ -0,0 +1,337 @@
+
+	/**
+	 * The current state domain at the currently processed AST node.
+	 */
+	private currentDomain: StateAbstractDomain<Domain>;
+
+	constructor(config: Config) {
+		super({ ...config, defaultVisitingOrder: 'forward', defaultVisitingType: 'exit' });
+
+		this.currentDomain = config.domain.bottom();
+	}
+
+	/**
+	 * Resolves the inferred abstract value of an AST node.
+	 * This requires that the abstract interpretation visitor has been completed, or at least started.
+	 * @param id     - The ID of the node to get the inferred value for
+	 * @param domain - An optional state abstract domain used to resolve the inferred abstract value (defaults to the state at the requested node)
+	 * @returns The inferred abstract value of the node, or `undefined` if no value was inferred for the node
+	 */
+	public getValue(id: RNode<ParentInformation & OtherInfo> | NodeId | undefined, domain?: StateAbstractDomain<Domain>): Domain | undefined {
+		const node = (id === undefined || typeof id === 'object') ? id : this.getNormalizedAst(id);
+		domain ??= node !== undefined ? this.getState(node.info.id) : undefined;
+
+		if(node === undefined) {
+			return;
+		} else if(domain?.has(node.info.id)) {
+			return domain.get(node.info.id);
+		}
+		const vertex = this.getDataflowGraph(node.info.id);
+		const call = vertex?.tag === VertexType.FunctionCall ? vertex : undefined;
+		const origins = Array.isArray(call?.origin) ? call.origin : [];
+
+		if(node.type === RType.Symbol) {
+			const values = this.getVariableOrigins(node.info.id).map(origin => domain?.get(origin));
+
+			if(values.length > 0 && values.every(isNotUndefined)) {
+				return AbstractDomain.joinAll(values);
+			}
+		} else if(node.type === RType.Argument && node.value !== undefined) {
+			return this.getValue(node.value, domain);
+		} else if(node.type === RType.ExpressionList && node.children.length > 0) {
+			return this.getValue(node.children[node.children.length - 1], domain);
+		} else if(node.type === RType.Pipe) {
+			return this.getValue(node.rhs, domain);
+		} else if(origins.includes('builtin:pipe')) {
+			if(node.type === RType.BinaryOp) {
+				return this.getValue(node.rhs, domain);
+			} else if(call?.args.length === 2 && call?.args[1] !== EmptyArgument) {
+				return this.getValue(call.args[1].nodeId, domain);
+			}
+		} else if(node.type === RType.IfThenElse) {
+			if(node.otherwise !== undefined) {
+				const values = [node.then, node.otherwise].map(entry => this.getValue(entry, domain));
+
+				if(values.length > 0 && values.every(isNotUndefined)) {
+					return AbstractDomain.joinAll(values);
+				}
+			}
+		} else if(origins.includes('builtin:if-then-else') && call?.args.every(arg => arg !== EmptyArgument)) {
+			if(call.args.length === 3) {
+				const values = call.args.slice(1, 3).map(entry => this.getValue(entry.nodeId, domain));
+
+				if(values.length > 0 && values.every(isNotUndefined)) {
+					return AbstractDomain.joinAll(values);
+				}
+			}
+		}
+	}
```

Comment:
Maybe we should discuss this, this should be abstract into a generic argument retriever based on the config, that is also to be used by others, we now have this code partially mirrored in various locations (besides your abstract semantic stuffsies afterward)